src/encoding.rs

```diff
@@ -5,15 +5,285 @@
+    pub fn import_bytes(&mut self, data: &[u8]) -> Result<(), postcard::Error> {
+        let update = postcard::from_bytes(data)?;
+        self.import(update);
+        Ok(())
+    }
+
+    /// Creates an iterator over the decoded metric items.
+    ///
+    /// Returns an iterator that yields [`Item`] instances combining schema and value data.
+    pub fn iter(&self) -> DecoderIter {
+        DecoderIter {
+            pos: 0,
+            inner: self,
+        }
+    }
+}
+
+/// Iterator over decoded metric items.
+///
+/// Iterates through the metrics in a [`Decoder`], yielding [`Item`] instances.
+#[derive(Debug)]
+pub struct DecoderIter<'a> {
+    /// Current position in the iteration
+    pos: usize,
+    /// Reference to the decoder being iterated
+    inner: &'a Decoder,
+}
+
+impl<'a> Iterator for DecoderIter<'a> {
+    type Item = Item<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let schema = self.inner.schema.as_ref()?.items.get(self.pos)?;
+        let value = self.inner.values.items.get(self.pos)?;
+        self.pos += 1;
+        Some(Item { schema, value })
+    }
+}
+
+impl MetricsSource for Decoder {
+    fn encode_openmetrics(&self, writer: &mut impl std::fmt::Write) -> Result<(), crate::Error> {
+        for item in self.iter() {
+            item.encode_openmetrics(writer)?;
+        }
+        write_eof(writer)?;
+        Ok(())
+    }
+}
+
+impl MetricsSource for Arc<RwLock<Decoder>> {
+    fn encode_openmetrics(&self, writer: &mut impl std::fmt::Write) -> Result<(), crate::Error> {
+        self.read().expect("poisoned").encode_openmetrics(writer)
+    }
+}
+
+/// Encoder for converting metrics from a registry into serializable updates.
+///
+/// Tracks schema changes and generates [`Update`] objects that can be
+/// transmitted to a [`Decoder`].
+#[derive(Debug)]
+pub struct Encoder {
+    /// The metrics registry to encode from
+    registry: Arc<RwLock<Registry>>,
+    /// Version of the last schema that was exported
+    last_schema_version: u64,
+}
+
+impl Encoder {
+    /// Creates a new encoder for the given registry.
+    ///
+    /// The encoder will track schema changes and only include schema
+    /// information in updates when it has changed.
+    pub fn new(registry: RwLockRegistry) -> Self {
+        Self {
+            registry,
+            last_schema_version: 0,
+        }
+    }
+
+    /// Exports the current state of the registry as an update.
+    ///
+    /// Returns an [`Update`] containing the current metric values and
+    /// optionally the schema (if it has changed since the last export).
+    pub fn export(&mut self) -> Update {
+        let registry = self.registry.read().expect("poisoned");
+        let current = registry.schema_version();
+        let schema = if current != self.last_schema_version {
+            self.last_schema_version = current;
+            let mut schema = Schema::default();
+            registry.encode_schema(&mut schema);
+            Some(schema)
```

Comment:
I guess I'll find out in review of downstream code, but what do we expect metrics receivers to do with conflicting descriptions? If I send a metric `bytes_sent` with a description of `"the number of bytes sent"`, then change the description to something else, will the receiver just last-writer-wins the new description?